inference/enterprise/workflows/execution_engine/compiler/graph_constructor.py

```diff
@@ -0,0 +1,602 @@
+    steps to execute.
+    Let's imagine graph:
+              / -> B -> C -> D
+    A -> IF <             \
+              \ -> E -> F -> G -> H
+    where node G requires node C even though IF branched the execution. In other
+    words - the problem (1) emerges if a node of kind STEP has a parent (node from which
+    it can be achieved) of kind STEP and this parent is in a different branch (point out that
+    we allow for a single step to have multiple steps as input, but they must be at the same
+    execution path - for instance if D requires an output from C and B - this is allowed).
+    Additionally, we must prevent situation when outcomes of branches started by two or more
+    condition steps merge with each other, as condition eval may result in contradictory
+    execution (2).
+
+
+    We need to detect that situations upfront, such that we can raise error of ambiguous execution path
+    rather than run time-consuming computations that will end up in error.
+
+    To detect problem (1), first we detect steps with more than one parent step.
+    From those steps we trace what sequence of steps would lead to execution engine reaching them.
+    We analyse paths from those parent nodes in reversed topological order (from those nodes towards entry
+    nodes of execution graph). While our analysis, on each path we denote control flow steps and
+    result of evaluation that must have been observed in runtime (which next step in normal flow must be
+    chosen to form the path). If we detect that for any control flow step we would need to output multiple
+    values at time (more than one registered next step of control flow step) - we raise error.
+
+    To detect problem (2) - we only let number of all different condition steps spotted in our traversal of
+    execution graph (while checking (1)) to be the max number condition steps in a single path from graph start node
+    to parent of problematic step with >= 2 parents.
+
+    Beware that the latter part of algorithm has quite bad time complexity in general case.
+    Worst part of algorithm runs at O(V^4) - at least taking coarse, worst-case estimations.
+    In fact, for reasonable workflows we can expect this is not so bad:
+    * The number of step nodes with multiple parents that we loop over in main loop, reduces the number of
+    steps we iterate through in inner loops, as we are dealing with DAG (with quite limited amount of edges)
+    and for each multi-parent node takes at least two other nodes (to construct a suspicious group) -
+    so expected number of iterations in main loop is low - let's say 1-3 for a real graph.
+    * for any reasonable execution graph, the complexity should be acceptable.
+    """
+    steps_with_more_than_one_parent = detect_steps_with_more_than_one_parent_step(
+        execution_graph=execution_graph
+    )  # O(V+E)
+    if len(steps_with_more_than_one_parent) == 0:
```

Comment:
`if len(steps_with_more_than_one_parent) == 0:` -> `if not steps_with_more_than_one_parent:`

Reply:
👍